File: recyclerlayout/src/main/java/com/angel/adapter/SWRecyclerAdapter.java
package com.angel.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

/**
 * Created by Angel on 2017/3/28.
 */
public abstract class SWRecyclerAdapter<T> extends RecyclerView.Adapter<SWViewHolder> {

    private List<T> list;
    private LayoutInflater inflater;

    public SWRecyclerAdapter(Context context, List<T> list) {
        this.list = list;
        inflater = LayoutInflater.from(context);
    }

    public SWViewHolder onCreateViewHolder(ViewGroup parent, int layoutID) {
        final SWViewHolder holder = new SWViewHolder(
                inflater.inflate(getItemLayoutId(layoutID), parent, false));
        return holder;
    }

    public void onBindViewHolder(SWViewHolder holder, int position) {
        bindData(holder, position, list.get(position));
    }

    abstract public int getItemLayoutId(int layoutID);

    abstract public void bindData(SWViewHolder holder, int position, T item);

    public int getItemCount() {
        return list.size();
    }

    public void add(int position, T item) {
        list.add(position, item);
        notifyItemInserted(position);
    }

    public void delete(int position) {
        list.remove(position);
        notifyItemRemoved(position);
    }

    public void clear() {
        if (list != null && list.size() > 0) {
            list.clear();
        }
        notifyDataSetChanged();
    }

    public void setTop(int position) {
        list.add(0, list.get(position));
        list.remove(position + 1);
        notifyDataSetChanged();
    }

}

File: kotlinlibrary/src/main/java/com/kotlin/adapter/SWRecyclerAdapter.kt
package com.kotlin.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

/**
 * Created by Angel on 2017/5/19.
 */
abstract class SWRecyclerAdapter<T>(context: Context, private val list: MutableList<T>?) : RecyclerView.Adapter<SWViewHolder>() {
    private val inflater: LayoutInflater

    init {
        inflater = LayoutInflater.from(context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, layoutID: Int): SWViewHolder {
        val holder = SWViewHolder(
                inflater.inflate(getItemLayoutId(layoutID), parent, false))
        return holder
    }

    override fun onBindViewHolder(holder: SWViewHolder, position: Int) {
        bindData(holder, position, list!![position])
    }

    abstract fun getItemLayoutId(layoutID: Int): Int

    abstract fun bindData(holder: SWViewHolder, position: Int, item: T)

    override fun getItemCount(): Int {
        return list!!.size
    }

    fun add(position: Int, item: T) {
        list!!.add(position, item)
        notifyItemInserted(position)
    }

    fun delete(position: Int) {
        list!!.removeAt(position)
        notifyItemRemoved(position)
    }

    fun clear() {
        if (list != null && list.size > 0) {
            list.clear()
        }
        notifyDataSetChanged()
    }

    fun setTop(position: Int) {
        list!!.add(0, list[position])
        list.removeAt(position + 1)
        notifyDataSetChanged()
    }

}
